src/metal/rt64_metal.cpp

```diff
@@ -0,0 +1,2363 @@
+        }
+
+        cmdBuffer->commit();
+
+        // Commit all command lists except the last one
+
+        for (uint32_t i = 0; i < commandListCount - 1; i++) {
+            assert(commandLists[i] != nullptr);
+
+            const MetalCommandList *interfaceCommandList = static_cast<const MetalCommandList *>(commandLists[i]);
+            MetalCommandList *mutableCommandList = const_cast<MetalCommandList*>(interfaceCommandList);
+            mutableCommandList->mtl->enqueue();
+            mutableCommandList->commit();
+        }
+
+        // Use the last command list to mark the end and signal the fence
+        const MetalCommandList *interfaceCommandList = static_cast<const MetalCommandList *>(commandLists[commandListCount - 1]);
+        interfaceCommandList->mtl->enqueue();
+
+        if (signalFence != nullptr) {
+            interfaceCommandList->mtl->addCompletedHandler([signalFence](MTL::CommandBuffer* cmdBuffer) {
+                dispatch_semaphore_signal(static_cast<MetalCommandFence *>(signalFence)->semaphore);
+            });
+        }
+
+        for (uint32_t i = 0; i < signalSemaphoreCount; i++) {
+            const MetalCommandSemaphore *interfaceSemaphore = static_cast<MetalCommandSemaphore *>(signalSemaphores[i]);
+            interfaceCommandList->mtl->encodeSignalEvent(interfaceSemaphore->mtl, interfaceSemaphore->mtlEventValue);
+        }
+
+        MetalCommandList *mutableCommandList = const_cast<MetalCommandList*>(interfaceCommandList);
+        mutableCommandList->commit();
+    }
+
+    void MetalCommandQueue::waitForCommandFence(RenderCommandFence *fence) {
+        const MetalCommandFence *metalFence = static_cast<MetalCommandFence *>(fence);
+        dispatch_semaphore_wait(metalFence->semaphore, DISPATCH_TIME_FOREVER);
+    }
+
+    // MetalPipelineLayout
+
+    MetalPipelineLayout::MetalPipelineLayout(MetalDevice *device, const RenderPipelineLayoutDesc &desc) {
+        assert(device != nullptr);
+
+        this->setLayoutCount = desc.descriptorSetDescsCount;
+
+        pushConstantRanges.resize(desc.pushConstantRangesCount);
+        memcpy(pushConstantRanges.data(), desc.pushConstantRanges, sizeof(RenderPushConstantRange) * desc.pushConstantRangesCount);
+    }
+
+    MetalPipelineLayout::~MetalPipelineLayout() {}
+
+    void MetalPipelineLayout::bindDescriptorSets(MTL::CommandEncoder* encoder, const std::unordered_map<uint32_t, MetalDescriptorSet*>& descriptorSets, const bool isCompute) const {
+        for (uint32_t i = 0; i < setLayoutCount; i++) {
+            if (descriptorSets.count(i) != 0) {
+                MetalDescriptorSet *descriptorSet = descriptorSets.at(i);
+                const MetalArgumentBuffer descriptorBuffer = descriptorSet->argumentBuffer;
+
+                //                descriptorBuffer.mtl->didModifyRange({0, DESCRIPTOR_RING_BUFFER_SIZE});
+
+                for (const auto& [descriptorIndex, resource] : descriptorSet->resources) {
```

Comment:
This is a predictable workload that we should just used a fixed size vector prepared ahead of time with what it needs to do. Iterating on a map will be significantly slower.